backend/algorithms/objects/conditions.py

```diff
@@ -211,78 +215,17 @@ def set_logic(self, logic):
+    def validate(self, user):
+        '''A helper function to be called by validate. The purpose of separating
         them is for easy warning implementation'''
-        # Ensure we add all the warnings. 
-        # NOTE: Remember that warnings are only returned
-        # along with True by validate() method. In other words, checking a warning 
-        # is != None is the equivalent of checking that validate() returned True
-        
         if self.conditions == []:
-            # Empty condition returns True by default
-            return True
-        
-        if self.logic == AND:
-            satisfied = True
-        else:
-            satisfied = False
-
-        for cond in self.conditions:
-            if isinstance(cond, (GRADECondition, WAMCondition)):
-                # Special type of condition which can return warnings
-                unlocked, warning = cond.validate(user)
-
-                if warning != None:
-                    warnings.append(warning)
-            elif isinstance(cond, CompositeCondition):
-                # Need to pass in the warnings list to collate all the warnings
-                unlocked = cond.validate(user, warnings)
-            else:
-                unlocked = cond.validate(user)
-
-            if self.logic == AND:
-                satisfied = satisfied and unlocked
-            else:
-                satisfied = satisfied or unlocked
-
-        return satisfied
-
-
-class FirstCompositeCondition(CompositeCondition):
-    '''The highest level composite condition (the outermost one). This is given
-    special treatment as this is the "entry point" to our algorithm'''
-    def __init__(self, course=None, logic=AND):
-        # The course which this condition applies to. Default value is None for testing purposes
-        self.course = course 
-        super().__init__()
-
-    def is_unlocked(self, user) -> dict:
-        '''The highest level check which returns the result and a warning. Call this
-        with the appropriate user data to determine if a course is unlocked or not.
-        Will return an object containing the result and a list of warnings'''
-        warnings = []
-
-        if self.course is not None:
-            result = {
-                "result": False,
-                "warnings": warnings
-            }
-            for exclusion in CACHED_EXCLUSIONS[self.course].keys():
-
-                if is_course(exclusion) and user.has_taken_course(exclusion):
-                    return result
-                elif is_program(exclusion) and user.in_program(exclusion):
-                    return result
-                else:
-                    # Not able to parse this type of  exclusion
-                    continue
-
-        unlocked = self.validate(user, warnings)
+            return True, []
 
-        return {
-            "result": unlocked,
-            "warnings": warnings
-        }
+        validations = [cond.validate(user) for cond in self.conditions]
```

Comment:
It's late at night but I'm just gonna trust you handled the warnings in a better way. The original purpose of FirstCompositeConditions was also to be the top level where warnings are collected. However seems like the team is coming up with something better!

Reply:
Basically what I did is that I forced every validate function to return a "warnings" list. This list is then flattened for composite conditions. I think this is better because it lets us deal with all "Condition" objects in the same way, allowing us to verify the return type and the fact that it has a "validate" method statically.

In a sense I basically used hindsight to unify all the conditions objects under one interface.